backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/tournament/TournamentService.java

```diff
@@ -0,0 +1,282 @@
+        User user = userRepository.findById(userId).orElseThrow(() -> new TutorException(USER_NOT_FOUND, userId));
+        CourseExecution courseExecution = courseExecutionRepository.findById(executionId).orElseThrow(() -> new TutorException(COURSE_EXECUTION_NOT_FOUND, executionId));
+
+        Set<Topic> topics = new HashSet<>();
+        for (Integer topicId : topicsId) {
+            Topic topic = topicRepository.findById(topicId)
+                    .orElseThrow(() -> new TutorException(TOPIC_NOT_FOUND, topicId));
+
+            topics.add(topic);
+        }
+
+        if (topics.isEmpty()) {
+            throw new TutorException(TOURNAMENT_MISSING_TOPICS);
+        }
+
+        Tournament tournament = new Tournament(user, courseExecution, topics, tournamentDto);
+        tournamentRepository.save(tournament);
+
+        return new TournamentDto(tournament);
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getAllTournamentsForCourseExecution(Integer executionId) {
+        return tournamentRepository.getAllTournamentsForCourseExecution(executionId).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getOpenedTournamentsForCourseExecution(Integer executionId) {
+        return tournamentRepository.getOpenedTournamentsForCourseExecution(executionId).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getClosedTournamentsForCourseExecution(Integer executionId) {
+        return tournamentRepository.getClosedTournamentsForCourseExecution(executionId).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getTournamentsByUserId(Integer userId) {
+        return tournamentRepository.getTournamentsByUserId(userId).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void joinTournament(Integer userId, TournamentDto tournamentDto, String password) {
+        User user = userRepository.findById(userId).orElseThrow(() -> new TutorException(USER_NOT_FOUND, userId));
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (DateHandler.now().isAfter(tournament.getEndTime())) {
```

Comment:
this check can be inside the addParticipant method

Reply:
Done